### PR TITLE
feat: dynamic batching

### PR DESCRIPTION
**Pull request by JohannesMessner**
allow dynamic batching for executors. Feature described in detail [here](https://github.com/jina-ai/jina/issues/5033).

## ToDo

- [x] create the decorator
- [ ] make overridable in YAML
- [ ] implement queue and batching in the worker
- [ ] implement waiting logic in the gateway
- [ ] documentation
- [ ] tests
- [ ] **Question:** what about mis-matching tensor shapes? let the executors handle it? let the decorator define an expected shape? Do smart batching based on the shape? (I like the last one)

### CODE REVIEW

**Comment by JoanFM** [comment]
> allow dynamic batching for executors. Feature described in detail [here](https://github.com/jina-ai/jina/issues/5033).
> 
> ## ToDo
> * [x]  create the decorator
> * [ ]  make overridable in YAML
> * [ ]  implement queue and batching in the worker
> * [ ]  implement waiting logic in the gateway
> * [ ]  documentation
> * [ ]  tests
> * [ ]  **Question:** what about mis-matching tensor shapes? let the executors handle it? let the decorator define an expected shape? Do smart batching based on the shape? (I like the last one)

As per the question, in our case batching is done at DocArray level so decotator does nothing with respect to the tensor for now.

A question from me: Have we researched the Triton behavior?

**Comment by JoanFM** [comment]
Make overridable in YAML means that there will be `runtime_args` that will make it overridable? I would suggest that we make it overridable even without the `decorator` so that we do not need to modify Hub Executors to use this behavior

File: jina/serve/runtimes/request_handlers/worker_request_handler.py [hunk 1]
```diff
@@ -266,6 +266,10 @@ async def handle(
         :param tracing_context: Optional OpenTelemetry tracing context from the originating request.
         :returns: the processed message
         """
+        # JOAN: This true? Why list?
+        # TODO: Just leave this then run all the tests
+        assert len(requests) == 1, "There are places where the handler is called with more than one request __apparently__"
```
Comment (JoanFM) [inline jina/serve/runtimes/request_handlers/worker_request_handler.py:271]: Sometimes if u have a Flow like this u will see:

```python
from jina import Flow, DocumentArray
with Flow().add(name='1').add(name='2', needs=['gateway']).add(name='3', needs=['1', '2'], no_reduce=True) as f:
   f.post(on='/', inputs=DocumentArray.empty(3))
```

**Comment by JohannesMessner** [comment]
> Make overridable in YAML means that there will be `runtime_args` that will make it overridable? I would suggest that we make it overridable even without the `decorator` so that we do not need to modify Hub Executors to use this behavior

definitely, it should work even if there is no decorator.

**Pull request closed without merging**